File: duplex/ts/src/record.test.ts
import { strict as assert } from 'node:assert';
import { readFileSync } from 'node:fs';
import { test } from 'node:test';
import { at, mount, type Message, type Path, type Wire } from './wire.ts';
import { record, MemoryWireLog, RecordError, type WireRecord } from './record.ts';

function deferred<T = void>() {
  let resolve!: (value: T) => void;
  const promise = new Promise<T>((r) => {
    resolve = r;
  });
  return { promise, resolve };
}
class Target implements Wire {
  refuse: Error | undefined;
  values: WireRecord[] = [];
  closed = false;
  changed = deferred();
  send(path: Path, message: Message) {
    if (this.refuse) throw this.refuse;
    if (this.closed) throw new Error('closed');
    this.values.push({ sequence: 0, path, message });
    this.changed.resolve();
    this.changed = deferred();
  }
  receive() {
    return () => {};
  }
  close() {
    this.closed = true;
  }
  async take() {
    while (!this.values.length) await this.changed.promise;
    return this.values.shift()!;
  }
}
const message = (value: number): Message => ({ frame: { version: 1, kind: 'event', data: value } });
class HeldLog extends MemoryWireLog {
  entered = deferred();
  release = deferred();
  holdAppend = false;
  failure: Error | undefined;
  async hold(signal: AbortSignal) {
    this.entered.resolve();
    await Promise.race([
      this.release.promise,
      new Promise<never>((_, reject) => {
        if (signal.aborted) reject(signal.reason);
        else signal.addEventListener('abort', () => reject(signal.reason), { once: true });
      }),
    ]);
    if (this.failure) throw this.failure;
  }
  override async read(sequence: number, signal: AbortSignal) {
    if (!this.holdAppend) await this.hold(signal);
    return super.read(sequence, signal);
  }
  override async append(path: Path, m: Message, signal: AbortSignal) {
    if (this.holdAppend) await this.hold(signal);
    return super.append(path, m, signal);
  }
}
const table = JSON.parse(
  readFileSync(new URL('../../../conformance/tables/recorded-wire.json', import.meta.url), 'utf8'),
) as {
  cases: Array<{ name: string; before: number[]; during: number[]; after: number; head: number; expected: number[] }>;
};
for (const row of table.cases)
  test(`record/follow: ${row.name}`, async () => {
    const store = new HeldLog();
    const w = await record(new Target(), store, { maxQueuedMessages: 8 });
    try {
      const source = at(mount(new Map([['history', w]])), ['history']);
      for (const v of row.before) source.send(['tick'], message(v));
      const target = new Target();
      const f = await w.follow(row.after, target);
      assert.equal(f.head, row.head);
      await store.entered.promise;
      for (const v of row.during) source.send(['tick'], message(v));
      assert.equal(await w.head(), row.before.length + row.during.length);
      store.release.resolve();
      const got = [];
      for (const _ of row.expected) {
        const entry = await target.take();
        assert.deepEqual(entry.path, ['tick']);
        assert.equal(entry.message.frame.kind, 'event');
        if (entry.message.frame.kind === 'event') got.push(entry.message.frame.data);
      }
      assert.deepEqual(got, row.expected);
      w.send(['fence'], message(99));
      const fence = await target.take();
      assert.deepEqual(fence.path, ['fence']);
      f.close();
      await f.done;
    } finally {
      w.close();
    }
  });
class BrokenSequenceLog extends MemoryWireLog {
  appendGap = false;
  override async append(path: Path, m: Message, signal: AbortSignal) {
    const next = await super.append(path, m, signal);
    return next + (this.appendGap ? 1 : 0);
  }
  override async read(at: number, signal: AbortSignal) {
    return { ...(await super.read(at, signal)), sequence: at + 1 };
  }
}
for (const mode of ['append', 'sequence', 'target'])
  test(`recorder ${mode} failure ends only its carrier and permits reentrant close observation`, async () => {
    const sentinel = new Error('consumer failure');
    const target = new Target();
    let store: MemoryWireLog = new MemoryWireLog();
    if (mode === 'append') {
      const held = new HeldLog();
      held.holdAppend = true;
      held.failure = sentinel;
      held.release.resolve();
      store = held;
    }
    if (mode === 'sequence') {
      const broken = new BrokenSequenceLog();
      broken.appendGap = true;
      store = broken;
    }
    if (mode === 'target') target.refuse = sentinel;
    const ended = deferred<unknown>();
    const w = await record(target, store, {
      onClose: (error) => {
        w.close();
        void assert.rejects(w.head()).then(() => ended.resolve(error));
      },
    });
    w.send([], message(1));
    const error = await ended.promise;
    if (mode === 'sequence') assert.ok(error instanceof RecordError && error.code === 'sequence');
    else assert.equal(error, sentinel);
    assert.equal(target.closed, true);
  });
for (const mode of ['read', 'sequence', 'target'])
  test(`follower ${mode} failure leaves recorder usable`, async () => {
    const sentinel = new Error('replay failure');
    let store: MemoryWireLog = new MemoryWireLog();
    if (mode === 'read') {
      const held = new HeldLog();
      held.failure = sentinel;
      held.release.resolve();
      store = held;
    }
    if (mode === 'sequence') store = new BrokenSequenceLog();
    await store.append([], message(1), new AbortController().signal);
    const w = await record(new Target(), store);
    try {
      const target = new Target();
      if (mode === 'target') target.refuse = sentinel;
      const f = await w.follow(0, target);
      await f.done;
      if (mode === 'sequence') assert.ok(f.error instanceof RecordError && f.error.code === 'sequence');
      else assert.equal(f.error, sentinel);
      assert.equal(await w.head(), 1);
    } finally {
      w.close();
    }
  });
test('stalled follower closes its mount while producer and healthy follower continue', async () => {
  const store = new HeldLog();
  const w = await record(new Target(), store, { maxQueuedMessages: 2 });
  try {
    w.send(['tick'], message(1));
    const slowTarget = new Target();
    const slow = await w.follow(0, at(mount(new Map([['slow', slowTarget]])), ['slow']));
    await store.entered.promise;
    const target = new Target();
    const fast = await w.follow(1, target);
    for (let i = 2; i <= 4; i++) {
      w.send(['tick'], message(i));
      await target.take();
    }
    await slow.done;
    assert.ok(slow.error instanceof RecordError);
    assert.equal(slow.error.code, 'overflow');
    assert.equal(slowTarget.closed, false);
    slowTarget.send(['probe'], message(99));
    w.send(['tick'], message(5));
    await target.take();
    fast.close();
    await fast.done;
  } finally {
    w.close();
  }
});
test('blocked storage cannot pace send beyond its admission bound', async () => {
  const store = new HeldLog();
  store.holdAppend = true;
  const target = new Target();
  const ended = deferred<unknown>();
  const w = await record(target, store, { maxQueuedMessages: 2, onClose: (error) => ended.resolve(error) });
  w.send([], message(1));
  await store.entered.promise;
  w.send([], message(2));
  w.send([], message(3));
  assert.throws(
    () => w.send([], message(4)),
    (error) => error instanceof RecordError && error.code === 'overflow',
  );
  assert.ok((await ended.promise) instanceof RecordError);
  assert.equal(target.closed, true);
});
test('cursor validation, cancellation and local return identity', async () => {
  const target = new Target();
  const store = new MemoryWireLog();
  const w = await record(target, store);
  try {
    await assert.rejects(
      w.follow(1, new Target()),
      (error) => error instanceof RecordError && error.code === 'sequence',
    );
    const address = { wire: target };
    w.send(['', 'é'], { ...message(1), return: address });
    const destination = new Target();
    const controller = new AbortController();
    const f = await w.follow(0, destination, controller.signal);
    const entry = await destination.take();
    assert.equal(entry.message.return, address);
    controller.abort();
    await f.done;
    assert.equal(f.error, controller.signal.reason);
    f.close();
    f.close();
    await assert.rejects(store.read(0, new AbortController().signal), RecordError);
  } finally {
    w.close();
  }
});

File: duplex/go/record_test.go
package duplex_test

import (
	"context"
	"encoding/json"
	"errors"
	bitwire "github.com/Bitspark/bitwire/wire/go"
	"os"
	"reflect"
	"sync"
	"testing"
	"time"

	"github.com/Bitspark/nightseam/duplex/go"
)

type recordTarget struct {
	values chan duplex.WireRecord
	closed chan struct{}
	once   sync.Once
	refuse error
}

func newRecordTarget() *recordTarget {
	return &recordTarget{values: make(chan duplex.WireRecord, 64), closed: make(chan struct{})}
}
func (w *recordTarget) Send(p []string, m bitwire.Message) error {
	if w.refuse != nil {
		return w.refuse
	}
	select {
	case <-w.closed:
		return duplex.ErrClosed
	default:
	}
	w.values <- duplex.WireRecord{Path: p, Message: m}
	return nil
}

type brokenSequenceLog struct {
	*duplex.MemoryWireLog
	appendGap bool
}

func (l *brokenSequenceLog) Append(ctx context.Context, p []string, m bitwire.Message) (uint64, error) {
	seq, err := l.MemoryWireLog.Append(ctx, p, m)
	if l.appendGap {
		seq++
	}
	return seq, err
}
func (l *brokenSequenceLog) Read(ctx context.Context, seq uint64) (duplex.WireRecord, error) {
	entry, err := l.MemoryWireLog.Read(ctx, seq)
	entry.Sequence++
	return entry, err
}

func TestRecordedWireStorageAndTargetFailures(t *testing.T) {
	for _, mode := range []string{"append", "sequence", "target"} {
		t.Run(mode, func(t *testing.T) {
			ctx := recordContext(t)
			sentinel := errors.New("consumer failure")
			target := newRecordTarget()
			var store duplex.WireLog = duplex.NewMemoryWireLog()
			if mode == "append" {
				held := newHeldLog()
				held.holdAppend = true
				held.failure = sentinel
				close(held.release)
				store = held
			}
			if mode == "sequence" {
				store = &brokenSequenceLog{MemoryWireLog: duplex.NewMemoryWireLog(), appendGap: true}
			}
			if mode == "target" {
				target.refuse = sentinel
			}
			ended := make(chan error, 1)
			var w *duplex.RecordedWire
			w, err := duplex.Record(ctx, target, store, duplex.RecordOptions{OnClose: func(err error) {
				_ = w.Close(1000, "reentrant")
				_, headErr := w.Head(ctx)
				if !errors.Is(headErr, duplex.ErrClosed) {
					ended <- headErr
					return
				}
				ended <- err
			}})
			if err != nil {
				t.Fatal(err)
			}
			if err := w.Send(nil, recordMessage(1)); err != nil {
				t.Fatal(err)
			}
			err = recordWait(t, ctx, ended)
			want := sentinel
			if mode == "sequence" {
				want = duplex.ErrRecordSequence
			}
			if !errors.Is(err, want) {
				t.Fatalf("got %v want %v", err, want)
			}
			select {
			case <-target.closed:
				t.Fatal("recorder closed its borrowed target")
			default:
			}
		})
	}
	for _, mode := range []string{"read", "sequence", "target"} {
		t.Run("follower "+mode, func(t *testing.T) {
			ctx := recordContext(t)
			sentinel := errors.New("replay failure")
			memory := duplex.NewMemoryWireLog()
			memory.Append(ctx, nil, recordMessage(1))
			var store duplex.WireLog = memory
			if mode == "read" {
				held := newHeldLog()
				held.MemoryWireLog = memory
				held.failure = sentinel
				close(held.release)
				store = held
			}
			if mode == "sequence" {
				store = &brokenSequenceLog{MemoryWireLog: memory}
			}
			w, err := duplex.Record(ctx, newRecordTarget(), store, duplex.RecordOptions{})
			if err != nil {
				t.Fatal(err)
			}
			defer w.Close(1000, "done")
			target := newRecordTarget()
			if mode == "target" {
				target.refuse = sentinel
			}
			f, err := w.Follow(ctx, 0, target)
			if err != nil {
				t.Fatal(err)
			}
			recordWait(t, ctx, f.Done)
			want := sentinel
			if mode == "sequence" {
				want = duplex.ErrRecordSequence
			}
			if !errors.Is(f.Err(), want) {
				t.Fatalf("got %v want %v", f.Err(), want)
			}
			if head, err := w.Head(ctx); err != nil || head != 1 {
				t.Fatalf("recorder affected: %d %v", head, err)
			}
		})
	}
}
func (*recordTarget) Receive(bitwire.Receiver) (func(), error) { return func() {}, nil }
func (w *recordTarget) Close(bitwire.Code, string) error {
	w.once.Do(func() { close(w.closed) })
	return nil
}
func recordMessage(i int) bitwire.Message {
	b, _ := json.Marshal(i)
	return bitwire.Message{Frame: bitwire.ProfileFrame{Version: 1, Kind: bitwire.ProfileEvent, Data: b}}
}
func recordContext(t *testing.T) context.Context {
	t.Helper()
	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
	t.Cleanup(cancel)
	return ctx
}
func recordWait[T any](t *testing.T, ctx context.Context, c <-chan T) T {
	t.Helper()
	select {
	case v := <-c:
		return v
	case <-ctx.Done():
		t.Fatal(ctx.Err())
		var z T
		return z
	}
}

type heldLog struct {
	*duplex.MemoryWireLog
	entered    chan struct{}
	release    chan struct{}
	once       sync.Once
	holdAppend bool
	failure    error
}

func newHeldLog() *heldLog {
	return &heldLog{MemoryWireLog: duplex.NewMemoryWireLog(), entered: make(chan struct{}), release: make(chan struct{})}
}

type heldHeadLog struct {
	*duplex.MemoryWireLog
	entered chan struct{}
}

func (l *heldHeadLog) Head(ctx context.Context) (uint64, error) {
	close(l.entered)
	<-ctx.Done()
	return 0, ctx.Err()
}
func TestRecordedWireSetupCanBeCancelled(t *testing.T) {
	ctx := recordContext(t)
	setup, cancel := context.WithCancel(ctx)
	store := &heldHeadLog{MemoryWireLog: duplex.NewMemoryWireLog(), entered: make(chan struct{})}
	target := newRecordTarget()
	done := make(chan error, 1)
	go func() { _, err := duplex.Record(setup, target, store, duplex.RecordOptions{}); done <- err }()
	recordWait(t, ctx, store.entered)
	cancel()
	if err := recordWait(t, ctx, done); !errors.Is(err, context.Canceled) {
		t.Fatal(err)
	}
	if err := target.Send(nil, recordMessage(1)); err != nil {
		t.Fatal("failed setup took ownership of target", err)
	}
}
func (l *heldLog) hold(ctx context.Context) error {
	l.once.Do(func() { close(l.entered) })
	select {
	case <-l.release:
		return l.failure
	case <-ctx.Done():
		return ctx.Err()
	}
}
func (l *heldLog) Read(ctx context.Context, seq uint64) (duplex.WireRecord, error) {
	if !l.holdAppend {
		if err := l.hold(ctx); err != nil {
			return duplex.WireRecord{}, err
		}
	}
	return l.MemoryWireLog.Read(ctx, seq)
}
func (l *heldLog) Append(ctx context.Context, p []string, m bitwire.Message) (uint64, error) {
	if l.holdAppend {
		if err := l.hold(ctx); err != nil {
			return 0, err
		}
	}
	return l.MemoryWireLog.Append(ctx, p, m)
}

func TestRecordedWireHeadAndHandoff(t *testing.T) {
	var table struct {
		Cases []struct {
			Name           string
			Before, During []int
			After, Head    uint64
			Expected       []int
		}
	}
	b, err := os.ReadFile("../../conformance/tables/recorded-wire.json")
	if err != nil {
		t.Fatal(err)
	}
	if err = json.Unmarshal(b, &table); err != nil {
		t.Fatal(err)
	}
	for _, c := range table.Cases {
		t.Run(c.Name, func(t *testing.T) {
			ctx := recordContext(t)
			store := newHeldLog()
			original := newRecordTarget()
			mounted := duplex.Mount(map[string]bitwire.Endpoint{"history": original})
			defer mounted.Close(1000, "done")
			w, err := duplex.Record(ctx, duplex.At(mounted, []string{"history"}), store, duplex.RecordOptions{MaxQueuedMessages: 8})
			if err != nil {
				t.Fatal(err)
			}
			defer w.Close(1000, "done")
			source := duplex.At(w, nil)
			for _, v := range c.Before {
				if err := source.Send([]string{"tick"}, recordMessage(v)); err != nil {
					t.Fatal(err)
				}
			}
			target := newRecordTarget()
			f, err := w.Follow(ctx, c.After, target)
			if err != nil {
				t.Fatal(err)
			}
			defer f.Close()
			if f.Head != c.Head {
				t.Fatalf("head %d", f.Head)
			}
			recordWait(t, ctx, store.entered)
			for _, v := range c.During {
				if err := source.Send([]string{"tick"}, recordMessage(v)); err != nil {
					t.Fatal(err)
				}
			}
			head, err := w.Head(ctx)
			if err != nil || head != uint64(len(c.Before)+len(c.During)) {
				t.Fatalf("producer cannot advance: %d %v", head, err)
			}
			close(store.release)
			var got []int
			for range c.Expected {
				e := recordWait(t, ctx, target.values)
				if !reflect.DeepEqual(e.Path, []string{"tick"}) {
					t.Fatal(e.Path)
				}
				var v int
				json.Unmarshal(e.Message.Frame.Data, &v)
				got = append(got, v)
			}
			if !reflect.DeepEqual(got, c.Expected) {
				t.Fatalf("got %v want %v", got, c.Expected)
			}
			if err := w.Send([]string{"fence"}, recordMessage(99)); err != nil {
				t.Fatal(err)
			}
			e := recordWait(t, ctx, target.values)
			if string(e.Message.Frame.Data) != "99" {
				t.Fatalf("duplicate before fence: %s", e.Message.Frame.Data)
			}
		})
	}
}

func TestRecordedWireStalledFollowerIsIsolated(t *testing.T) {
	ctx := recordContext(t)
	store := newHeldLog()
	original := newRecordTarget()
	w, err := duplex.Record(ctx, original, store, duplex.RecordOptions{MaxQueuedMessages: 2})
	if err != nil {
		t.Fatal(err)
	}
	defer w.Close(1000, "done")
	w.Send([]string{"tick"}, recordMessage(1))
	slowTarget := newRecordTarget()
	mounted := duplex.Mount(map[string]bitwire.Endpoint{"slow": slowTarget})
	defer mounted.Close(1000, "done")
	slow, err := w.Follow(ctx, 0, duplex.At(mounted, []string{"slow"}))
	if err != nil {
		t.Fatal(err)
	}
	recordWait(t, ctx, store.entered)
	fastTarget := newRecordTarget()
	fast, err := w.Follow(ctx, 1, fastTarget)
	if err != nil {
		t.Fatal(err)
	}
	defer fast.Close()
	for i := 2; i <= 4; i++ {
		if err := w.Send([]string{"tick"}, recordMessage(i)); err != nil {
			t.Fatal(err)
		}
		recordWait(t, ctx, fastTarget.values)
	}
	recordWait(t, ctx, slow.Done)
	if !errors.Is(slow.Err(), duplex.ErrRecordOverflow) {
		t.Fatal(slow.Err())
	}
	// The follower borrows send authority; both mount and target stay usable.
	if err := slowTarget.Send([]string{"probe"}, recordMessage(99)); err != nil {
		t.Fatal(err)
	}
	if err := w.Send([]string{"tick"}, recordMessage(5)); err != nil {
		t.Fatal(err)
	}
	recordWait(t, ctx, fastTarget.values)
}

func TestRecordedWireSlowStorageBoundsAdmission(t *testing.T) {
	ctx := recordContext(t)
	store := newHeldLog()
	store.holdAppend = true
	target := newRecordTarget()
	ended := make(chan error, 1)
	w, err := duplex.Record(ctx, target, store, duplex.RecordOptions{MaxQueuedMessages: 2, OnClose: func(err error) { ended <- err }})
	if err != nil {
		t.Fatal(err)
	}
	defer w.Close(1000, "done")
	w.Send(nil, recordMessage(1))
	recordWait(t, ctx, store.entered)
	w.Send(nil, recordMessage(2))
	w.Send(nil, recordMessage(3))
	if err := w.Send(nil, recordMessage(4)); !errors.Is(err, duplex.ErrRecordOverflow) {
		t.Fatal(err)
	}
	if err := recordWait(t, ctx, ended); !errors.Is(err, duplex.ErrRecordOverflow) {
		t.Fatal(err)
	}
	select {
	case <-target.closed:
		t.Fatal("recorder closed its borrowed target")
	default:
	}
}

func TestRecordedWireCursorCancellationAndCapabilities(t *testing.T) {
	ctx := recordContext(t)
	store := duplex.NewMemoryWireLog()
	target := newRecordTarget()
	w, err := duplex.Record(ctx, target, store, duplex.RecordOptions{})
	if err != nil {
		t.Fatal(err)
	}
	defer w.Close(1000, "done")
	if _, err := w.Follow(ctx, 1, newRecordTarget()); !errors.Is(err, duplex.ErrRecordSequence) {
		t.Fatal(err)
	}
	address := &bitwire.ReturnAddress{Wire: target}
	message := recordMessage(1)
	message.Return = address
	w.Send([]string{"", "é"}, message)
	followerTarget := newRecordTarget()
	life, cancel := context.WithCancel(ctx)
	f, err := w.Follow(life, 0, followerTarget)
	if err != nil {
		t.Fatal(err)
	}
	e := recordWait(t, ctx, followerTarget.values)
	if e.Message.Return != address {
		t.Fatal("return capability changed")
	}
	cancel()
	recordWait(t, ctx, f.Done)
	if !errors.Is(f.Err(), context.Canceled) {
		t.Fatal(f.Err())
	}
	f.Close()
	f.Close()
	if _, err := store.Read(ctx, 0); !errors.Is(err, duplex.ErrRecordSequence) {
		t.Fatal(err)
	}
}
